netconf_client/connect.py

```diff
@@ -7,6 +7,30 @@
 from netconf_client.log import logger
 
 
+def __connect(host, port, timeout):
+    """
+    Try to connect to the all addresses.
+    https://docs.python.org/3/library/socket.html
+    """
+    s = None
+    for res in socket.getaddrinfo(host, port, socket.AF_UNSPEC, socket.SOCK_STREAM):
+        af, socktype, proto, canonname, sa = res
+        try:
+            s = socket.socket(af, socktype, proto)
+        except OSError as msg:
+            s = None
+            continue
+        try:
+            s.settimeout(timeout)
+            s.connect(sa)
+        except OSError as msg:
+            s.close()
+            s = None
+            continue
+        break
+    return s
```

Comment:
This should probably raise an `OSError` if `s` is still `None`, to be consistent with the old behavior

Reply:
It would be a timeout error, due to line 24
`           s.settimeout(timeout)`

Reply:
I'm not sure I follow.

In the old behavior, if the socket timed out or otherwise failed to connect, the `connect_ssh` function would raise an `OSError` of some kind.

If any of these function calls fail, some flavor of `OSError` (like e.g. `socket.timeout`) will be raised but immediately caught by lines 20 and 26. This means that if all connections fail, this function will return `None` instead of raising an exception.

What this code probably needs to do is to save off the most recent exception in a variable, and re-raise that exception if all connections fail.

Reply:
fixed: raise Exception if not connected